Add support for Google Gemini API integration and enhance LLM provide…

Uzair-akrum opened the pull request:
Implemented new functions for managing Gemini API key and model 

Updated IPC handlers to support Gemini API interactions

Modified settings UI to allow selection between OpenAI and Gemini providers
 
![Screenshot from 2025-06-29 01-21-15](https://github.com/user-attachments/assets/86d083c3-a673-4f99-960c-35c6d2b6db84)


![Screenshot from 2025-06-29 01-20-55](https://github.com/user-attachments/assets/21f95253-8575-4e91-965f-da87a4fd233f)

Uzair-akrum committed:
Add support for Google Gemini API integration and enhance LLM provider settings

NickTikhonov commented:
Thank you for the change! I'll test and provide feedback over the weekend

NickTikhonov commented on package.json:
```diff
@@ -25,6 +25,7 @@
     "@codemirror/lang-sql": "^6.9.0",
     "@electron-toolkit/preload": "^3.0.1",
     "@electron-toolkit/utils": "^4.0.0",
+    "@google/generative-ai": "^0.24.1",
```
it looks like this package was deprecated:

https://www.npmjs.com/package/@google/generative-ai

NickTikhonov commented on src/main/lib/llm/gemini.ts:
```diff
@@ -0,0 +1,77 @@
+      }
+    })
+    
+    const systemPrompt = `You are a SQL (postgres) and data visualization expert. Your job is to help the user write or modify a SQL query to retrieve the data they need.
```
Weird to have the system prompt hidden inside a model adapter.

NickTikhonov commented on src/main/lib/llm/index.ts:
```diff
@@ -0,0 +1,41 @@
+  return adapter
+}
+
+export function clearLLMCache() {
```
I don't understand why this cache is needed

NickTikhonov commented on src/main/lib/llm/openai.ts:
```diff
@@ -0,0 +1,52 @@
+
+  async generateText({ prompt, model = 'gpt-4o', temperature, maxTokens }: LLMGenOptions): Promise<string> {
+    // Check if this is a SQL query generation prompt (contains schema information)
+    if (prompt.includes('table schema') || prompt.includes('SQL') || prompt.includes('postgres')) {
```
This is going to result in a plethora of bugs and unexpected behavior in the future

Uzair-akrum committed:
Merge remote-tracking branch 'upstream/main' into main

Uzair-akrum committed:
Merge remote-tracking branch 'upstream/main' into main

Uzair-akrum committed:
Merge remote-tracking branch 'upstream/main' into main

Uzair-akrum committed:
- Refactored LLM adapters
- Removed  keyword guessing logic from the OpenAI adapter.
- Centralized all prompt construction logic into the
`generateQuery` IPC handler in `main/index.ts`
- Removed the unnecessary in-memory cache

Uzair-akrum committed:
Update package dependencies and refactor GeminiAdapter to use new Google GenAI SDK

Uzair-akrum commented:
Thanks for the detailed feedback @NickTikhonov 
i have removed the keyword guessing logic in openai adaptor 
The prompt construction part is now centralized in main/index.ts.
Cache logic removed and deprecated package is also replaced

Uzair-akrum committed:
Merge remote-tracking branch 'upstream/main' into main

Uzair-akrum committed:
Updated Gemini and OpenAI adapters to support new QueryResponse type, ensuring consistent data format across LLMs

Uzair-akrum committed:
Refactor index.ts and GeminiAdapter to return full QueryResponse object, ensuring consistent data handling across the application

Uzair-akrum commented:
Hello @NickTikhonov I've updated the LLM adapters type definitions to produce and handle the new QueryResponse object and with the responseMimeType Gemini Adaptor is perfectly in sync with the zod  schema that is being used in openai adaptor

NickTikhonov commented on src/main/index.ts:
```diff
@@ -106,26 +113,84 @@ app.whenReady().then(() => {
+    await setLLMProvider(provider)
+  })
+
+  ipcMain.handle('generateWithLLM', async (_, provider, prompt, opts) => {
```
Why is this needed?

NickTikhonov commented on src/main/lib/db.ts:
```diff
@@ -1,8 +1,5 @@
 import pg from 'pg'
```
I don't like that you're moving this specific logic out of this file and into index.ts - in general there are a lot of arbitrary refactors in this PR that I don't agree with, rather than simply focusing on implementing the main change

NickTikhonov reviewed (changes requested):
This PR seems really overcomplicated for what it's trying to accomplish. 

All it has to do is import this package and use it as an adapter instead of `createOpenAI`

https://ai-sdk.dev/providers/ai-sdk-providers/google-generative-ai 

Doing this will drop the PR to about 1/3 of the size and make it overall much cleaner

NickTikhonov commented:
Check out how I've done it with the anthropic integration. You should be able to extend it pretty easily

Uzair-akrum committed:
Merge remote-tracking branch 'upstream/main' into main

Uzair-akrum committed:
Merge remote-tracking branch 'upstream/main' into main

Uzair-akrum committed:
Add support for Gemini AI provider and update dependencies

Uzair-akrum committed:
Remove OpenAI adapter and related types from the LLM module, updating references in preload index.d.ts

Uzair-akrum commented:
@NickTikhonov you integrating claude made a lot of things easier for me to understand what needs to change
I have refrained from doing any additional refactors as in the previous adapter pattern i was enforced to do so
I have made relevant changes to add Gemini support reducing pr size to 3x
